### ISSUE TITLE
this.requestAnimationFrame is not a function appears when going in and out of fullscreen mode

### ISSUE THREAD

**Issue by bubah** [GeekyAnts/NativeBase#3206]
## I have gone through these following points

- [x] Check latest documentation: https://docs.nativebase.io/
- [x] Check for existing open/closed issues for a possible duplicate before creating a new issue: https://github.com/GeekyAnts/NativeBase/issues
- [x] Use the latest NativeBase release: https://github.com/GeekyAnts/NativeBase/releases
- [x] Check examples from NativeBase KitchenSink https://github.com/GeekyAnts/NativeBase-KitchenSink
- [] For discussion purpose make use of NativeBase Slack: http://slack.nativebase.io/
- [] For queries related to theme, check [Theme Variables](https://docs.nativebase.io/docs/ThemeVariables.html) from Docs and live NativeBase Theme Editor http://nativebase.io/customizer/

## Issue Description
When playing a video in our app and we go in and out of fullscreen mode the simulator throws the error ```this.requestAnimationFrame is not a function```
When using a prior version of native-base (2.13.12) this bug did not exist in our app. Once we upgraded to native-base version 2.13.13, the bug started to appear. 

We've forked the Native-base repo and implemented the following workaround in the node_modules/native-base/src/basic/Tabs/index.js file:

<img width="960" alt="Screen Shot 2020-07-29 at 3 09 21 PM" src="https://user-images.githubusercontent.com/15990942/88842568-ac63f200-d1ad-11ea-92cc-b303e297fa91.png">

And in the node_modules/native-base/dist/src/basic/Tabs/index.js file:

<img width="1525" alt="Screen Shot 2020-07-29 at 3 14 27 PM" src="https://user-images.githubusercontent.com/15990942/88843064-7ecb7880-d1ae-11ea-9963-72e4f6dec014.png">

#### node, npm, react-native, react and native-base version, expo version if used, xcode version
node: 10.15.3
react-native: 0.60.0
react: 16.8.6
native-base: 2.13.13
xcode: 11.1 (11A1027)

#### Expected behaviour
Should not throw an error when going in and out of fullscreen mode

#### Actual behaviour
throws an error when going in and out of fullscreen mode
<img width="469" alt="Screen Shot 2020-07-29 at 2 58 57 PM" src="https://user-images.githubusercontent.com/15990942/88841695-4d51ad80-d1ac-11ea-9c45-35013160bd02.png">

#### Steps to reproduce
Launch app and play a video
Then click fullscreen button to play video in fullscreen mode

#### Is the bug present in both iOS and Android or in any one of them?
Yes, bug is present in iOS and Android.

#### Any other additional info which would help us debug the issue quicker.

###Important

If you want your issue to be looked at quicker, attach a snack reproducible with your issue. Makes it easier for us!

**Comment by Babazon** [comment]
We have to fork and fix it ourselves too. A PR would take a month to be reviewed and merge on the library. 
We will go back to installing from npm once it's fixed.

**Comment by sunviwo** [comment]
Please @Babazon , could you tell me how you fixed it? Facing the same issue for multiple hours now.

**Comment by ealmiladi** [comment]
> We have to fork and fix it ourselves too. A PR would take a month to be reviewed and merge on the library.
> We will go back to installing from npm once it's fixed.

Yes, please post a solution for this.

**Comment by Babazon** [comment]
Guys I did the same thing as the original post. Read it. 

Fork the library to your organisation/user 

Go to native-base/src/basic/Tabs/index.js file:
Find "this.requestAnimationFrame(. " invocation, and add before it this.requestAnimationFrame &&.  (If this function is undefined, the AND gate && will not allow the rest of the expression to be executed)

Go to /native-base/dist/src/basic/Tabs/index.js file:
Find the same function and add the same expression before it 

Add your changes, commit and push to your own NativeBAse repo 

Install by installing from the repository you forked to, instead of npm

`yarn add github:YOUR_GIT_ORGANISATION_OR_USER_NAME/NativeBase`

This will cause issues switching landscape and portrait orientation in the view you had the tabs.

**Comment by droM4X** [comment]
> Go to /native-base/dist/src/basic/Tabs/index.js file:
Find the same function and add the same expression before it

This is not neccessary, just edit the code in src directory. 
Use the `npm run prepublish` command, this will compile your code to dist, than commit it to your fork.

**Comment by shobujroy** [comment]
> Guys I did the same thing as the original post. Read it.
> 
> Fork the library to your organisation/user
> 
> Go to native-base/src/basic/Tabs/index.js file:
> Find "this.requestAnimationFrame(. " invocation, and add before it this.requestAnimationFrame &&. (If this function is undefined, the AND gate && will not allow the rest of the expression to be executed)
> 
> Go to /native-base/dist/src/basic/Tabs/index.js file:
> Find the same function and add the same expression before it
> 
> Add your changes, commit and push to your own NativeBAse repo
> 
> Install by installing from the repository you forked to, instead of npm
> 
> `yarn add github:YOUR_GIT_ORGANISATION_OR_USER_NAME/NativeBase`
> 
> This will cause issues switching landscape and portrait orientation in the view you had the tabs.

Can you please help me? I'm new to react-native. Here is the code of that file.

`  _handleLayout(e) {
    const { width } = e.nativeEvent.layout;

    if (
      !width ||
      width <= 0 ||
      Math.round(width) === Math.round(this.state.containerWidth)
    ) {
      return;
    }
    this.setState({ containerWidth: width });
    this.requestAnimationFrame( '' && () => {
      this.goToPage(this.state.currentPage);
    });
  },`

**Comment by ealmiladi** [comment]
> > Guys I did the same thing as the original post. Read it.
> > Fork the library to your organisation/user
> > Go to native-base/src/basic/Tabs/index.js file:
> > Find "this.requestAnimationFrame(. " invocation, and add before it this.requestAnimationFrame &&. (If this function is undefined, the AND gate && will not allow the rest of the expression to be executed)
> > Go to /native-base/dist/src/basic/Tabs/index.js file:
> > Find the same function and add the same expression before it
> > Add your changes, commit and push to your own NativeBAse repo
> > Install by installing from the repository you forked to, instead of npm
> > `yarn add github:YOUR_GIT_ORGANISATION_OR_USER_NAME/NativeBase`
> > This will cause issues switching landscape and portrait orientation in the view you had the tabs.
> 
> Can you please help me? I'm new to react-native. Here is the code of that file.
> 
> ` _handleLayout(e) {
> const { width } = e.nativeEvent.layout;
> 
> ```
> if (
>   !width ||
>   width <= 0 ||
>   Math.round(width) === Math.round(this.state.containerWidth)
> ) {
>   return;
> }
> this.setState({ containerWidth: width });
> this.requestAnimationFrame( () => {
>   this.goToPage(this.state.currentPage);
> });
> ```
> 
> },`

```
this.requestAnimationFrame && this.requestAnimationFrame( () => {
   this.goToPage(this.state.currentPage);
 });
```

**Comment by shobujroy** [comment]
> > > Guys I did the same thing as the original post. Read it.
> > > Fork the library to your organisation/user
> > > Go to native-base/src/basic/Tabs/index.js file:
> > > Find "this.requestAnimationFrame(. " invocation, and add before it this.requestAnimationFrame &&. (If this function is undefined, the AND gate && will not allow the rest of the expression to be executed)
> > > Go to /native-base/dist/src/basic/Tabs/index.js file:
> > > Find the same function and add the same expression before it
> > > Add your changes, commit and push to your own NativeBAse repo
> > > Install by installing from the repository you forked to, instead of npm
> > > `yarn add github:YOUR_GIT_ORGANISATION_OR_USER_NAME/NativeBase`
> > > This will cause issues switching landscape and portrait orientation in the view you had the tabs.
> > 
> > 
> > Can you please help me? I'm new to react-native. Here is the code of that file.
> > ` _handleLayout(e) {
> > const { width } = e.nativeEvent.layout;
> > ```
> > if (
> >   !width ||
> >   width <= 0 ||
> >   Math.round(width) === Math.round(this.state.containerWidth)
> > ) {
> >   return;
> > }
> > this.setState({ containerWidth: width });
> > this.requestAnimationFrame( () => {
> >   this.goToPage(this.state.currentPage);
> > });
> > ```
> > 
> > 
> > },`
> 
> ```
> this.requestAnimationFrame && this.requestAnimationFrame( () => {
>    this.goToPage(this.state.currentPage);
>  });
> ```
> 
> Should follow what @droM4X said

Yes, I did it and it's fixed noq but the tab transitions still very slow. Is there anything that still I need to fix?

**Comment by ikroeber** [comment]
Same issue here when using Tabs component inside a Modal.

**Comment by huylvdev** [comment]
same issues, I reinstalled version 2.13.8

**Comment by jaferi512** [comment]
Waiting For Solution
My Environment:
React Native: 0.63.2
Native Base: 2.13.13

**Comment by Xopsy** [comment]
Same issue. I had to downgrade to 2.13.12 and fix in my node modules some issues with Toast and Item because of useNativeDriver is required. And on Tabs this issue when you rotate your device.

**Comment by dgeorgiev** [comment]
Hey, pushed a PR with a fix for this https://github.com/GeekyAnts/NativeBase/pull/3218

**Comment by tejparekh** [comment]
Hello I am facing the same issues and I couldn't find a solution on this thread or the other threads mentioned in the comments here. Does the solution involve manually changing the file in the package? 

I have upgraded from the 0.57 to 0.63 in steps and now I am able to compile the code but I am getting the Run time errors. 
![Screenshot 2020-09-14 at 09 49 41](https://user-images.githubusercontent.com/13714772/93031620-ad72b500-f66f-11ea-9e8a-71fea2f46f4d.png)

**Current setup of core with native base.** 

 "native-base": "^2.13.14",
    "react": "16.13.1",
    "react-emoji-render": "^1.2.4",
    "react-moment": "^0.9.7",
    "react-native": "0.63.2",

Thanks for the help. Appreciate getting some response soon.

**Comment by elilambnz** [comment]
@tejparekh if you don't want to mess around with manually changing files, you could **temporarily** import @dgeorgiev's handy PR into npm in place of `native-base`:
`"native-base": "Healthyco/NativeBase#feature/fix-request-animation"`

Or `npm i`/`yarn add` `Healthyco/NativeBase#feature/fix-request-animation`

Just remember to subscribe to PR #3218 changes and revert back to importing `native-base` once it's merged in.

**Comment by Quiec** [comment]
> @tejparekh if you don't want to mess around with manually changing files, you could **temporarily** import @dgeorgiev's handy PR into npm in place of `native-base`:
> `"native-base": "Healthyco/NativeBase#feature/fix-request-animation"`
> 
> Or `npm i`/`yarn add` `Healthyco/NativeBase#feature/fix-request-animation`
> 
> Just remember to subscribe to PR #3218 changes and revert back to importing `native-base` once it's merged in.

I installed this but didn't work! Can you explain with more details?

**Comment by elilambnz** [comment]
@Quiec npm can install a package directly from a github URL. In this case, I'm swapping out the usual way of installing the `native-base` package with a forked feature branch found [here](https://github.com/Healthyco/NativeBase/tree/feature/fix-request-animation) that patches the issue. You can read more about the ways to install npm packages [here](https://docs.npmjs.com/cli-commands/install.html).

If this doesn't answer your question, or the problem is not related to installing the patch, you could post any error messages you're getting. Alternatively, you could create a new issue on the appropriate repo and feel free to tag me in it.

**Comment by Quiec** [comment]
@elilambnz Nope, my error same and I know npm. But even though I run your code, the error is the same, it doesn't change.
In short, pull-request doesn't work (for me)

**Comment by Quiec** [comment]
@elilambnz I cleared cache & reinstalled; now working! Thanks!!!

**Comment by HeyShinner** [comment]
> Guys I did the same thing as the original post. Read it.
> 
> Fork the library to your organisation/user
> 
> Go to native-base/src/basic/Tabs/index.js file:
> Find "this.requestAnimationFrame(. " invocation, and add before it this.requestAnimationFrame &&. (If this function is undefined, the AND gate && will not allow the rest of the expression to be executed)
> 
> Go to /native-base/dist/src/basic/Tabs/index.js file:
> Find the same function and add the same expression before it
> 
> Add your changes, commit and push to your own NativeBAse repo
> 
> Install by installing from the repository you forked to, instead of npm
> 
> `yarn add github:YOUR_GIT_ORGANISATION_OR_USER_NAME/NativeBase`
> 
> This will cause issues switching landscape and portrait orientation in the view you had the tabs.

It works for me!

**Comment by comvenger-brandon** [comment]
Is this library dead / discontinued?

**Comment by Enalmada** [comment]
> Is this library dead / discontinued?

Latest build crashes all the time because of this and it has been that way for a long time now so I am looking for other options.  What is the next best thing for React Native?  Seems like these are the top ones I can find...

https://github.com/react-native-elements/react-native-elements
https://akveo.github.io/react-native-ui-kitten
https://callstack.github.io/react-native-paper

**Comment by elilambnz** [comment]
If you're after the tab navigation specifically, UI Kitten has a pretty nice [Tab Bar component](https://akveo.github.io/react-native-ui-kitten/docs/components/top-tabs/overview#tabbar). I quite like the look of that library overall and would probably choose it today if I were to replace NativeBase.

**Comment by elilambnz** [comment]
> @tejparekh if you don't want to mess around with manually changing files, you could **temporarily** import @dgeorgiev's handy PR into npm in place of `native-base`:
> `"native-base": "Healthyco/NativeBase#feature/fix-request-animation"`
> 
> Or `npm i`/`yarn add` `Healthyco/NativeBase#feature/fix-request-animation`
> 
> Just remember to subscribe to PR #3218 changes and revert back to importing `native-base` once it's merged in.

Just a reminder to everyone who used this workaround, PR #3218 has been merged in. You should be able to update your import for `native-base` by reinstalling through npm/yarn.

This issue should now be closed.

**Comment by ignacioToptive** [comment]
@elilambnz I've just checked and #3218 is still open, so there's no reason to close this issue yet

**Comment by elilambnz** [comment]
@ignacioToptive oh, you're right, it's still open. Out of curiosity, does the latest version of `native-base` solve this issue for you? I updated to the latest and it seems to be ok on iOS, but I haven't tested Android yet.

In any case, my apologies! Maybe hold off switching the import just yet if it doesn't solve the problem.

**Comment by ignacioToptive** [comment]
@elilambnz I'm on latest available version (2.13.14) and the bug is still present

**Comment by rdgoutiyama** [comment]
When change screen orientation crashes

**Comment by diogoaltoe** [comment]
I already have two corrections made in native-base.

I'm using the Patch Package library (https://github.com/ds300/patch-package) to make these corrections.

Below is the patch file with this "this.requestAnimationFrame" correction and another one related to the use of the Picker library.

[native-base+2.13.14.patch.zip](https://github.com/GeekyAnts/NativeBase/files/5451842/native-base%2B2.13.14.patch.zip)

**Comment by Andrerm124** [comment]
Can we expect an update to be released any time soon with this bugfix?
Kind of frustrating having a redbox every time I open certain screens

**Comment by janpaulalmanoche** [comment]
> > @tejparekh if you don't want to mess around with manually changing files, you could **temporarily** import @dgeorgiev's handy PR into npm in place of `native-base`:
> > `"native-base": "Healthyco/NativeBase#feature/fix-request-animation"`
> > Or `npm i`/`yarn add` `Healthyco/NativeBase#feature/fix-request-animation`
> > Just remember to subscribe to PR #3218 changes and revert back to importing `native-base` once it's merged in.
> 
> Just a reminder to everyone who used this workaround, PR #3218 has been merged in. You should be able to update your import for `native-base` by reinstalling through npm/yarn.
> 
> This issue should now be closed.

i did yarn add Healthyco/NativeBase#feature/fix-request-animation and it fixed my problem, im testing via android

**Comment by spaceboom73** [comment]
Okey, i'm solved this problem by comments above, all changes have in my fork https://github.com/spaceboom73/NativeBase
Here solved problem of importing Picker and problem with requestAnimationFrame error.
If you are lazy find solutions, use my fork repository with yarn add or npm i github:spaceboom73/NativeBase
It's solution is temporary, while NativeBase not release new version

**Comment by edodusi** [comment]
So v2.13.15 out and this is still an issue unfortunately. So maybe it's not a library bug but just the way we are using it?

**Comment by tyhour** [comment]
use this package, (yarn add https://github.com/tyhour/NativeBase.git) I just fixed some issue and all warning with Native Base. Also fixed error (this.requestAnimationFrame is not a function appears). I wish this can help you. :)

**Comment by edodusi** [comment]
@tyhour it's good that we can use forked packages with fixes (thanks for that) but the thing is, if the maintainers don't think this is an issue at all, in months, and they keep releasing new versions that still contains this "bug", maybe this is not a "bug" but just we are using this library the wrong way or maybe using incompatible packages, don't know.

It would be appreciated to get just some kind of official statement by the maintainers but I understand it's not easy dealing with all these issues here, this became a big project and we are using it for free so I totally understand, just wondering why are we getting this error and everybody else in the world is not...

**Comment by lcastillo88** [comment]
I created my own PR (https://github.com/GeekyAnts/NativeBase/pull/3316) based on native-base v2.13.14 which I am currently using. It only has the fix for the TypeError and nothing else. If you want to use it temporarily while the code owners patch/fix the bug or merge my PR, do this:

`"native-base": "lcastillo88/NativeBase"`
Or 
`npm i/yarn add lcastillo88/NativeBase`

**Comment by gino8080** [comment]
For me the problem is fixed in latest 2.15.2 version !

**Comment by majugurci** [comment]
For me this is still an issue using 2.15.2 on Android.

**Comment by HeyShinner** [comment]
For me the problem is fixed in v2.13.14
